### PR TITLE
FlightTask: Fix ekf2 reset race condition during task switch

### PR DESCRIPTION
**Pull request by bresch**
fixes https://github.com/PX4/Firmware/issues/14677

During a loss of GPS data when using GPS as primary height source, the height is reset to baro and the local position gets invalid at the same time. This triggers a switch to altitude flight task and a setpoint reset.
This combination of events had the effect to ignore the height reset, the large sudden height error could create an abrupt change of altitude or even a crash.
The ekf2 reset is now done at the beginning of each update call.

The ekf reset without a correct compensation of the setpoint can lead to a crash.
task_no_z_reset
task_z_reset

### CODE REVIEW

File: src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp [hunk 1]
```diff
@@ -60,6 +60,7 @@ void FlightTaskTransition::checkSetpoints(vehicle_local_position_setpoint_s &set
 
 bool FlightTaskTransition::update()
 {
+	FlightTask::update();
```
Comment (MaEtUgR) [inline src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp:63]: `FlightTaskOffboard` and `FlightTaskOrbit` need to call the base update as well.
Also to make it fully correct you need to logically and with the return value like for example here: https://github.com/PX4/Firmware/blob/2fca03e1bda313f86fd96d05130e830b9a2ee220/src/lib/flight_tasks/tasks/Manual/FlightTaskManual.cpp#L47

File: src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp [hunk 2]
```diff
@@ -37,17 +36,13 @@ bool FlightTask::updateInitialize()
...
 	_evaluateVehicleLocalPosition();
 	_evaluateDistanceToGround();
-	_checkEkfResetCounters();
 	return true;
 }
 
-void FlightTask::_initEkfResetCounters()
+bool FlightTask::update()
 {
-	_reset_counters.xy = _sub_vehicle_local_position.get().xy_reset_counter;
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp:46]: Can we be sure the reset counter is never zero when there's a valid estimate?

[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp:46]: Don't you then get an e.g. `_ekfResetHandlerPositionZ()` and `_smoothing.setCurrentPosition(_position(2))` after every task switch which would override taking over the setpoint from the last task?

[the remark this answers is not in the document]
Comment (bresch) [reply src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp:46]: > Don't you then get an e.g. _ekfResetHandlerPositionZ() and _smoothing.setCurrentPosition(_position(2)) after every task switch which would override taking over the setpoint from the last task?

Ouch, yes, this is correct and not something we want. I need to find an other solution, thanks...

[the remark this answers is not in the document]
Comment (bresch) [reply src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp:46]: Now, the previous value of the counters is transferred to from the previous to the next FlightTask using a get/set function

**Comment by bresch** [comment]
@MaEtUgR Should be good now. I tested in SITL that the reset works when EKF resets and that the counter is saved when you switch between tasks

File: src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp (hunk 1 above)
Comment (bresch) [reply src/lib/flight_tasks/tasks/Transition/FlightTaskTransition.cpp:63]: Thanks, I added it for `FlightTaskOffboard`, but orbit inherits from Altitude, so it's not required

File: src/lib/flight_tasks/FlightTasks.hpp [hunk 3]
```diff
@@ -81,6 +81,12 @@ class FlightTasks
 	 */
 	const vehicle_local_position_setpoint_s getPositionSetpoint();
 
+	/**
+	 * Get the reset counters from the current task
```
Comment (MaEtUgR) [inline src/lib/flight_tasks/FlightTasks.hpp:85]: The name says that but I would probably ask myself what reset counters.
```suggestion
	 * Get the estimation reference frame reset counters from the current task
```
?

File: src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp [hunk 4]
```diff
@@ -37,17 +37,13 @@ bool FlightTask::updateInitialize()
...
-	_reset_counters.vxy = _sub_vehicle_local_position.get().vxy_reset_counter;
-	_reset_counters.z = _sub_vehicle_local_position.get().z_reset_counter;
-	_reset_counters.vz = _sub_vehicle_local_position.get().vz_reset_counter;
-	_reset_counters.quat = _sub_attitude.get().quat_reset_counter;
+	_checkEkfResetCounters();
```
Comment (MaEtUgR) [inline src/lib/flight_tasks/tasks/FlightTask/FlightTask.cpp:45]: In principle you could leave the `_checkEkfResetCounters()` call in the update Initialize. They would just be checked twice on activation then... works either way.

File: src/lib/flight_tasks/tasks/FlightTask/FlightTask.hpp [hunk 5]
```diff
@@ -113,6 +121,9 @@ class FlightTask : public ModuleParams
 	 */
 	const vehicle_local_position_setpoint_s getPositionSetpoint();
 
+	const ekf_reset_counters_s getResetCounters() const { return _reset_counters; }
+	void setResetCounters(ekf_reset_counters_s counters) { _reset_counters = counters; }
```
Comment (MaEtUgR) [inline src/lib/flight_tasks/tasks/FlightTask/FlightTask.hpp:125]:
```suggestion
	void setResetCounters(const ekf_reset_counters_s &counters) { _reset_counters = counters; }
```

File: src/lib/flight_tasks/tasks/Offboard/FlightTaskOffboard.cpp [hunk 6]
```diff
@@ -153,7 +155,7 @@ bool FlightTaskOffboard::update()
 		_position_setpoint.setNaN(); // Don't require any position/velocity setpoints
 		_velocity_setpoint.setNaN();
 		_acceleration_setpoint = Vector3f(0.f, 0.f, 100.f); // High downwards acceleration to make sure there's no thrust
-		return true;
+		return ret;
```
Comment (MaEtUgR) [inline src/lib/flight_tasks/tasks/Offboard/FlightTaskOffboard.cpp:158]: omg we should to get rid of these early returns, not here.

Comment (bresch) [reply src/lib/flight_tasks/tasks/Offboard/FlightTaskOffboard.cpp:158]: Yes, it's horrible :|

**Pull request merged**